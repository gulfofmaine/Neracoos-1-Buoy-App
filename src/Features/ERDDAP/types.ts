/**
 * Types related to the ERDDAP service
 */

import { Feature } from "@turf/helpers"

import { ReadingTimeSeries } from "@app/Shared/timeSeries"

export interface ERDDAPStoreState {
  errorMessage?: string
  loading: boolean
  platforms: PlatformFeatureWithDatasets[]
  forecasts: ForecastStore
}

export interface ForecastStore {
  forecasts: ForecastSource[]
  loading: boolean
  errorMessage?: string
}

export const initialStoreState: ERDDAPStoreState = {
  forecasts: {
    forecasts: [],
    loading: false
  },
  loading: false,
  platforms: []
}

export interface PlatformFeatureCollection {
  features: PlatformFeature[]
}

export type PlatformFeature = Feature & {
  properties: PlatformProperties & {
    readings: PlatformTimeSeries[]
  }
}

export type PlatformFeatureWithDatasets = Feature & {
  properties: PlatformProperties & {
    readings: PlatformDataset[]
    forecasts: ForecastDataset[]
  }
}

export interface PlatformProperties {
  alerts: PlatformAlert[]
  attribution: PlatformAttribution[]
  mooring_site_desc: string
  nbdc_site_id?: string
  // uscg_light_letter?: string
  // watch_circle_radius?: number
}

export interface PlatformAlert {
  start_time: string
  end_time?: string
  message: string
  level: string
}

export interface PlatformTimeSeries {
  value?: number
  time?: string
  depth: number
  data_type: DataType
  server: string
  variable: string
  dataset: string
  start_time: string
  constraints: {
    [key: string]: string | number
  }
}

export type PlatformDataset = PlatformTimeSeries & {
  loading: boolean
  loadStartTimes: Date[]
  error: string
  readings: ReadingTimeSeries[]
}

export enum ForecastType {
  waveHeight = "Wave Height",
  wavePeriod = "Wave Period",
  waveDirection = "Wave Direction",
  airTemperature = "Air Temperature",
  windSpeed = "Wind Speed",
  windDirection = "Wind Direction"
}

export interface ForecastSource {
  slug: string
  forecast_type: string
  name: string
  description: string
  source_url: string
  point_forecast: string
}

export interface ForecastDataset {
  source: ForecastSource
  loading: boolean
  error: string
  readings: ReadingTimeSeries[]
}

export type ForecastJson = ForecastSource & {
  latitude: number
  longitude: number
  time_series: ReadingTimeSeries[]
}

export interface FetchGroup {
  server: string
  dataset: string
  constraints: {
    [key: string]: string | number
  }
  datasets: PlatformDataset[]
}

export interface PlatformAttribution {
  program: PlatformProgram
  attribution: string
}

export interface PlatformProgram {
  name: string
  website: string
}

export interface DataType {
  standard_name: string
  short_name: string
  long_name: string
  units: string
}
