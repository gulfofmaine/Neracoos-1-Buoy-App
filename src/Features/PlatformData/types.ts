import { ErddapDataset } from '@app/Shared/erddap'
import { ReadingTimeSeries } from '@app/Shared/timeSeries'

export enum Status {
    Loaded = 'LOADED',
    Loading = 'LOADING',
    Error = 'ERROR'
}

export interface PlatformData {
    data_type: string
    unit: string
    depth: number
    data: ReadingTimeSeries[]
}

export type ErddapDatasetInfo = ErddapDataset & {
    coverageStart: Date
    coverageEnd: Date
    status: Status
    error_message: string
}

export type DatasetData = PlatformData & ErddapDataset & {
    status: Status
    error_message: string
}

export interface Platform {
    id: string
    status: Status
    error_message: string
    data_types: PlatformData[]
    forecasts_types: DatasetData[]
}

interface PlatformJsonDataTypeDict {
    depth: string[]
    uom: string
}

interface PlatformJsonDataType {
    [props: string]: PlatformJsonDataTypeDict
}

export interface PlatformJsonTimeSeriesReading {
    depth: string
    quality: string
    reading: string
}

interface PlatformJsonTimeSeries {
    [props: string]: PlatformJsonTimeSeriesReading
}

export interface PlatformJsonBase {
    data_types: PlatformJsonDataType
    // date
    platform: string
}

export type PlatformJson = PlatformJsonBase & {
    [props: string]: PlatformJsonTimeSeries[]
}

export interface PlatformDataStoreState {
    datasetInfo: ErddapDatasetInfo[]
    platforms: Platform[]
}

export const initialStoreState: PlatformDataStoreState = {
    datasetInfo: [],
    platforms: []
}
