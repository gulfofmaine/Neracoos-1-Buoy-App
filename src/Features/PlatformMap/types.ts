import { Feature } from '@turf/helpers'

export interface PlatformLocationsJson {
    // http://www.neracoos.org/data/json/monitoringlocations.php
    [props: string]: PlatformLocation
}

export interface PlatformDataDepths {
    [props: string]: number[]
}

export interface PlatformDataTypes {
    [props: string]: string
}

export interface PlatformProperties {
    active: boolean
    program: string
    platform_credit: string
    mooring_site_desc: string
    name: string
}

export interface PlatformLocation {
    active: boolean
    data_depths: PlatformDataDepths
    data_types: PlatformDataTypes
    lat: number
    lon: number
    mooring_site_desc: string
    platform_credit: string
    program: string
}

export interface PlatformMapStoreState {
    platforms: Feature[]
}