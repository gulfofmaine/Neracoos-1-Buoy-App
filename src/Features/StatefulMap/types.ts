import type { RView } from "rlayers/RMap"

export interface MapState {
  view?: RView
}

export const initialStoreState: MapState = {
  view: undefined,
}
