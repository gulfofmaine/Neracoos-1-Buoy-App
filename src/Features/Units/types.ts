/**
 * Types related to unit managent
 */

export enum UnitSystem {
  metric = "Metric",
  imperial = "Imperial"
}

export interface UnitStoreState {
  system: UnitSystem
}

export const initialStoreState: UnitStoreState = {
  system: UnitSystem.imperial
}
