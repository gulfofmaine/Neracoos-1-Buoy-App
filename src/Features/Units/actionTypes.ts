/**
 * Unit system management action types
 */

export const UNIT_SWITCH = "unit/switch"
export type UNIT_SWITCH = typeof UNIT_SWITCH
