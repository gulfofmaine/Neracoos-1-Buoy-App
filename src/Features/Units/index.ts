export { unitReducer } from "./reducer"

export { UnitSelector } from "./UnitSelector"
