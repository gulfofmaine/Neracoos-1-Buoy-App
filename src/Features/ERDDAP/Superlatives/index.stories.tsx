import * as React from "react"

import { UnitSystem } from "Features/Units/types"

import { Superlatives, ShowSuperlatives } from "./index"

import { platforms } from "stories/platforms"

export default {
  component: Superlatives,
  title: "ERDDAP/Superlatives",
}

export const configureable = (args) => <ShowSuperlatives {...args} />
configureable.args = {
  unitSystem: UnitSystem.english,
  platforms,
  laterThan: new Date("2020-09-23T19:49:02.212Z"),
}

export const shouldNotShowOutdatedData = (args) => <ShowSuperlatives {...args} />
shouldNotShowOutdatedData.args = {
  unitSystem: UnitSystem.english,
  platforms,
  laterThan: new Date("2020-09-30T19:49:02.212Z"),
}
