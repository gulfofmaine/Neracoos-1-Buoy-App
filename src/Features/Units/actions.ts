import * as actionTypes from "./actionTypes"
import { UnitSystem } from "./types"

// Action types

export interface UnitSwitch {
  type: typeof actionTypes.UNIT_SWITCH
  system: UnitSystem
}

export type UnitActions = UnitSwitch

// Action creators

/**
 * Switch display to use designated system of units
 * @param system system of  units to switch display into
 */
export const unitSwitch = (system: UnitSystem): UnitSwitch => {
  return {
    system,
    type: actionTypes.UNIT_SWITCH,
  }
}
