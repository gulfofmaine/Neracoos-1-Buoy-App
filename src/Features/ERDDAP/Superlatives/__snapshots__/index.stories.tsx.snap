// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ERDDAP/Superlatives Configurable smoke-test 1`] = `
<div class="card"
     style="margin-top: 1rem; margin-bottom: 1rem;"
>
  <div class="card-header">
    <h5>
      Latest Conditions
    </h5>
  </div>
  <div class="card-body">
    <div class="row">
      <div class="col">
        <h6>
          Highest Winds
        </h6>
        <div>
          22.8 Knots
        </div>
        <a href="/platform/WLIS">
          <div>
            WLIS: Western Long Island Sound
          </div>
        </a>
      </div>
      <div class="col">
        <h6>
          Biggest Waves
        </h6>
        <div>
          8.3 Feet
        </div>
        <a href="/platform/44008">
          <div>
            44008: Southeast of Nantucket
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ERDDAP/Superlatives ShouldNotShowOutdatedData smoke-test 1`] = `
<div class="card"
     style="margin-top: 1rem; margin-bottom: 1rem;"
>
  <div class="card-header">
    <h5>
      Latest Conditions
    </h5>
  </div>
  <div class="card-body">
    <div class="row">
      <div class="col">
        <h6>
          Highest Winds
        </h6>
        <div>
          -- --
        </div>
      </div>
      <div class="col">
        <h6>
          Biggest Waves
        </h6>
        <div>
          -- --
        </div>
      </div>
    </div>
  </div>
</div>
`;
